runtime/lua/vim/filetype.lua

```diff
@@ -1724,522 +1728,614 @@ local filename = {
 }
 
 -- Re-use closures as much as possible
-local detect_apache = starsetf('apache')
-local detect_muttrc = starsetf('muttrc')
-local detect_neomuttrc = starsetf('neomuttrc')
+local detect_apache_diretc = starsetf('apache', { fastpat = '/etc/' })
+local detect_apache_dotconf = starsetf('apache', { fastpat = '%.conf' })
+local detect_muttrc = starsetf('muttrc', { fastpat = 'utt' })
+local detect_neomuttrc = starsetf('neomuttrc', { fastpat = 'utt' })
+local detect_xkb = starsetf('xkb', { fastpat = '/usr/' })
 
---- @type vim.filetype.mapping
+-- Table of filetype pattern matching rules grouped by their "fast pattern"
+--
+-- Every filetype pattern match is prefaced with a (cached) matching of its parent "fast pattern".
+-- If there is no match, skip all matching inside group. For that to work, fast pattern should:
+-- - Match at least the same set of strings as filetype patterns inside it. But not too much more.
+-- - Be fast to match (will be matched directly with `string.match(target, fastpat)`).
+--
+-- The core idea behind speedup is that matching results of fast patterns are cached and reused
+-- inside single same `M.match()` call. The effect is as big as the quality of compromise between:
+-- - Fast pattern groups many filetype patterns: leads to better cache reuse in case of no match.
+-- - Fast pattern is specific: leads to less "fast pattern matches but regular one does not" which
+--   has the penalty of a (not useful) single extra fast pattern match.
+--
+-- Example:
+-- - Regular pattern: '.*/etc/a2ps/.*%.cfg'.
+-- - Good fast patterns: '/etc/'; '%.cfg$' (depends on how many patterns they can group)
+-- - Bad fast patterns: '%.' (fast but not specific), '/a2ps/.*%.' (slow but specific)
+--
+-- When adding a new filetype with pattern matching rule, consider the following:
+-- - If there is already a group with appropriate fast pattern, use it.
+-- - If there can be a fast and specific enough pattern to group at least 3 filetype patterns,
+--   add it as a separate grouped entry. Good new fast pattern should be:
+--     - Fast. Good rule of thumb is that it should be a short specific string (i.e. no quantifiers
+--       or character sets).
+--     - Specific. Good rules of thumb (from best to worst):
+--         - Full directory name (like '/etc/', '/log/').
+--         - Part of a rare enough directory name (like '/conf', 'git/').
+--         - Character sequence reasonably rarely used in real full paths (like 'nginx', 'cmus').
+--
+-- When modifying an existing regular pattern, make sure that it still fits its group.
```

Comment:
This comment block can be significantly cut down.

Reply:
Which parts do you think are not necessary?

I tried to make it more compact and this is the result, as any information seems valuable to me. Can not trim down without more guidance.

Reply:
I've used [your suggestion](https://github.com/neovim/neovim/pull/29660#discussion_r1675665603). Thanks for trimming it down!